### ISSUE TITLE
Add group of mappings in main mapping section based on assigned virtual targets

### ISSUE THREAD

**Issue by jackmau** [helgoboss/realearn#189]
Last FR for the day, I promise ^^

In the old realearn world (v1) one would have to create a mapping, move controller, select target and so on. The process was quite tedious, as a lot of controller transmit more than one message, and you would often have to go and change each mapping individually. Now thanks to @helgoboss prolific work we have virtual mapping and we have to do that only once ..._in theory_... in practice once you create your main mappings you have to go and select per each mapping its virtual source from the dropdown menu, which, in my opinion, is hardly an improvement time-wise. Why just not add an entire set of assignments based on the mapped virtual target, so we only have to map targets? It's much quicker to delete unused mappings than to have to go in the dropdown menu to select the virtual targets. An alternative could be simple to be able to duplicate groups, as in most user case you would tend to have the same virtual controllers in each group.

**Comment by helgoboss** [comment]
Controller presets with virtual targets are not really intended as a tool to reduce mapping effort.

I think what you want is this: #161

**Comment by jackmau** [comment]
> Controller presets with virtual targets are not really intended as a tool to reduce mapping effort.

I see that, but I don't see why penalising the users of virtual mappings when normal mappings can do multiple mappings quickly whereas virtual sources are only accessible through dropdown with 3 clicks. In my mind (aside the amazing controller projection) the whole point of virtual mappings _is_ saving time as, quoting the manual, _You don't need to learn your control elements again and again_. Sorry for my frustration, but the main _selling point_ of realearn was the ability to learn multiple assignment quickly, I feel that has been lost with virtual mappings and I don't see any technical reason not to allow user to load a set of mappings with their virtual controller of choice and then doing multiple mappings just of the controlled targets.

> I think what you want is this: #161

I have read this, but I am not sure I understand it. I thought this related to entire mappings (both targets and sources), whereas I am only thinking to sources.

**Comment by helgoboss** [comment]
> > Controller presets with virtual targets are not really intended as a tool to reduce mapping effort.
> 
> I see that, but I don't see why penalising the users of virtual mappings when normal mappings can do multiple mappings quickly whereas virtual sources are only accessible through dropdown with 3 clicks. In my mind (aside the amazing controller projection) the whole point of virtual mappings _is_ saving time as, quoting the manual, _You don't need to learn your control elements again and again_. Sorry for my frustration, but the main _selling point_ of realearn was the ability to learn multiple assignment quickly, I feel that has been lost with virtual mappings and I don't see any technical reason not to allow user to load a set of mappings with their virtual controller of choice and then doing multiple mappings just of the controlled targets.
> 

Can you make a mockup or something, some little drawing of what you have in mind? And maybe a screencast of the process that is so annoying with the current way of doing things. At first I need to understand the problem. Probably you are using ReaLearn in a way which I never really had in mind. Seems like you want to have some kind of One-Source-to-Many-Targets thing? But then why would you map one source to so many targets? Would they all be active at the same time or being made active/inactive using conditional activation? Expressing requirements is not easy, I know, but it's the most important thing for an FR.

From my perspective, using virtual sources are as quick to use as MIDI sources: Press "Learn source", touch the control element and you are done. Using the menus to set sources is rather a fallback solution.

> > I think what you want is this: #161
> 
> I have read this, but I am not sure I understand it. I thought this related to entire mappings (both targets and sources), whereas I am only thinking to sources.

This is about inheriting parts from a kind of prototype mapping, where parts can be the whole source, a part of the source, the whole target, a part of the target etc. It's for easier management and reuse.

**Comment by jackmau** [comment]
> From my perspective, using virtual sources are as quick to use as MIDI sources: Press "Learn source", touch the control element and you are done. Using the menus to set sources is rather a fallback solution.

That was exactly my problem.  I am sorry @helgoboss, I did not know that that Realearn recognised automatically virtual sources. I did some more testing today and it is indeed my fault, Realearn (2.5) recognised correctly the CMC buttons as virtual inputs when learning sources for main mappings, however the same was not true for faders, which transmit note on on first touch (whereas I was interested in mapping the consequent CC). In practice as long as I start to move the fader before doing the assignments the fader is recognised successfully. It would be nice if one could set some messages to be ignored for those control surfaces that exhibit this behaviour, but I don't think there is the need to add a group of mappings with virtual sources input.
